File: src/main/java/com/learning/leetcode/patterns/Triangle.java
package com.learning.leetcode.patterns;

public class Triangle {
    public static void main(String[] args) {
        int n = 5;
        System.out.println("Triangle stars pattern");
        triangleStars(n);
        System.out.println("Triangle stars Reverse pattern");
        triangleStarsReverse(n);
        System.out.println("Triangle numbers pattern");
        triangleNumbers(n);
        System.out.println("Triangle same number row pattern");
        triangleNumbers2(n);
        System.out.println("Triangle numbers alternate pattern");
        triangleNumbers3(n);
        System.out.println("Triangle numbers incrementing pattern");
        triangleNumbers4(n);
        System.out.println("Triangle numbers symmetric pattern");
        triangleNumbersIncrement(n);
        System.out.println("Triangle numbers Reverse pattern");
        triangleNumbersReverse(n);
        System.out.println("Triangle characters pattern");
        triangleCharacters(n);
        System.out.println("Triangle alphabets pattern 2");
        triangleCharacters2(n);
        System.out.println("Triangle alphabets reverse pattern");
        triangleCharactersReverse(n);
        System.out.println("Triangle alphabets reverse pattern 2");
        triangleCharacters3(n);
    }

    private static void triangleStars(int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                System.out.print("* ");
            }
            System.out.println();
        }
    }

    private static void triangleStarsReverse(int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n - i; j++) {
                System.out.print("* ");
            }
            System.out.println();
        }
    }

    private static void triangleNumbers(int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                System.out.print((j + 1) + " ");
            }
            System.out.println();
        }
    }

    private static void triangleNumbers2(int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                System.out.print((i + 1) + " ");
            }
            System.out.println();
        }
    }

    private static void triangleNumbers3(int n) {
        for (int i = 0; i < n; i++) {
            int start = i % 2 == 0 ? 1 : 0;
            for (int j = 0; j <= i; j++) {
                System.out.print(start + " ");
                start = 1 - start;
            }
            System.out.println();
        }
    }

    private static void triangleNumbers4(int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i + 1; j++) {
                System.out.print(j + 1 + " ");
            }
            for (int j = 0; j < 2 * (n - i - 1); j++) {
                System.out.print("  ");
            }
            for (int j = i + 1; j > 0; j--) {
                System.out.print(j + " ");
            }
            System.out.println();
        }
    }

    private static void triangleNumbersIncrement(int n) {
        int count = 1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                System.out.print(count++ + " ");
            }
            System.out.println();
        }
    }

    private static void triangleNumbersReverse(int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n - i; j++) {
                System.out.print((j + 1) + " ");
            }
            System.out.println();
        }
    }

    private static void triangleCharacters(int n) {
        for (int i = 0; i < n; i++) {
            for (char j = 'A'; j <= 'A' + i; j++) {
                System.out.print(j + " ");
            }
            System.out.println();
        }
    }

    private static void triangleCharacters2(int n) {
        for (char i = 0; i < n; i++) {
            char ch = (char) ('A' + i);
            for (char j = 0; j <= i; j++) {
                System.out.print(ch + " ");
            }
            System.out.println();
        }
    }


    private static void triangleCharactersReverse(int n) {
        for (int i = 0; i < n; i++) {
            for (char j = 'A'; j < 'A' + n - i; j++) {
                System.out.print(j + " ");
            }
            System.out.println();
        }
    }

    private static void triangleCharacters3(int n) {
        for (int i = 0; i < n; i++) {
            for (char ch = (char) ('E' - i); ch <= 'E'; ch++) {
                System.out.print(ch + " ");
            }
            System.out.println();
        }
    }
}

File: src/main/java/com/learning/leetcode/patterns/Triangle.py
def triangleStars(n: int):
    for i in range(n):
        for j in range(i + 1):
            print("*", end=" ")
        print()


def triangleStarsReverse(n: int):
    for i in range(n):
        for j in range(n - i):
            print("*", end=" ")
        print()


def triangleNumbers(n: int):
    for i in range(n):
        for j in range(i + 1):
            print(str(j + 1), end=" ")
        print()


def triangleNumbers2(n: int):
    for i in range(n):
        for j in range(i + 1):
            print(str(i + 1), end=" ")
        print()


def triangleNumbers3(n: int):
    for i in range(n):
        start = 1 if i % 2 == 0 else 0
        for j in range(i + 1):
            print(str(start), end=" ")
            start = 1 - start
        print()


def triangleNumbers4(n: int):
    for i in range(n):
        for j in range(i + 1):
            print(str(j + 1), end=" ")
        for j in range(2 * (n - i - 1)):
            print(" ", end=" ")
        for j in range(i, -1, -1):
            print(str(j + 1), end=" ")
        print()


def triangleNumbersReverse(n: int):
    for i in range(n):
        for j in range(n - i):
            print(str(j + 1), end=" ")
        print()


def triangleNumbersIncrement(n: int):
    count = 1
    for i in range(n):
        for j in range(i + 1):
            print(str(count), end=" ")
            count += 1
        print()


def triangleCharacters(n: int):
    for i in range(n):
        for j in range(ord('A'), ord('A') + i + 1, 1):
            print(chr(j), end=" ")
        print()


def triangleCharacters2(n: int):
    for i in range(n):
        ch = ord('A') + i
        for j in range(i+1):
            print(chr(ch), end=" ")
        print()


def triangleAlphabetsReverse(n: int):
    for i in range(n):
        for j in range(ord('A'), ord('A') + n - i, 1):
            print(chr(j), end=" ")
        print()


def triangleAlphabetsReverse2(n: int):
    for i in range(n):
        start = ord('E') - i
        end = ord('E') + 1
        for j in range(start, end):
            print(chr(j), end=" ")
        print()


def main():
    print("Triangle star pattern")
    triangleStars(5)
    print("Triangle star reverse pattern")
    triangleStarsReverse(5)
    print("Triangle number pattern")
    triangleNumbers(5)
    print("Triangle same number row pattern")
    triangleNumbers2(5)
    print("Triangle alternate number pattern")
    triangleNumbers3(5)
    print("Symmetric triangle number pattern")
    triangleNumbers4(5)
    print("Triangle number reverse pattern")
    triangleNumbersReverse(5)
    print("Triangle number increment pattern")
    triangleNumbersIncrement(5)
    print("Triangle characters pattern")
    triangleCharacters(5)
    print("Triangle characters pattern 2")
    triangleCharacters2(5)
    print("Triangle alphabets reverse pattern")
    triangleAlphabetsReverse(5)
    print("Triangle alphabets reverse pattern 2")
    triangleAlphabetsReverse2(5)


main()
